samples/net/cloud/aws_iot_mqtt/src/main.c

```diff
@@ -0,0 +1,494 @@
+
+	client_ctx.rx_buf = rx_buffer;
+	client_ctx.rx_buf_size = MQTT_BUFFER_SIZE;
+	client_ctx.tx_buf = tx_buffer;
+	client_ctx.tx_buf_size = MQTT_BUFFER_SIZE;
+
+	/* setup TLS */
+	client_ctx.transport.type = MQTT_TRANSPORT_SECURE;
+	struct mqtt_sec_config *const tls_config = &client_ctx.transport.tls.config;
+
+	tls_config->peer_verify = TLS_PEER_VERIFY_REQUIRED;
+	tls_config->cipher_list = NULL;
+	tls_config->sec_tag_list = sec_tls_tags;
+	tls_config->sec_tag_count = ARRAY_SIZE(sec_tls_tags);
+	tls_config->hostname = CONFIG_AWS_ENDPOINT;
+	tls_config->cert_nocopy = TLS_CERT_NOCOPY_NONE;
+}
+
+struct backoff_context {
+	uint16_t retries_count;
+	uint16_t max_retries;
+
+#if defined(CONFIG_AWS_EXPONENTIAL_BACKOFF)
+	uint32_t attempt_max_backoff; /* ms */
+	uint32_t max_backoff;	      /* ms */
+#endif
+};
+
+static void backoff_context_init(struct backoff_context *bo)
+{
+	__ASSERT_NO_MSG(bo != NULL);
+
+	bo->retries_count = 0u;
+	bo->max_retries = MAX_RETRIES;
+
+#if defined(CONFIG_AWS_EXPONENTIAL_BACKOFF)
+	bo->attempt_max_backoff = BACKOFF_EXP_BASE_MS;
+	bo->max_backoff = BACKOFF_EXP_MAX_MS;
+#endif
+}
+
+/* https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/ */
+static void backoff_get_next(struct backoff_context *bo, uint32_t *next_backoff_ms)
+{
+	__ASSERT_NO_MSG(bo != NULL);
+	__ASSERT_NO_MSG(next_backoff_ms != NULL);
+
+#if defined(CONFIG_AWS_EXPONENTIAL_BACKOFF)
+	if (bo->retries_count <= bo->max_retries) {
+		*next_backoff_ms = sys_rand32_get() % (bo->attempt_max_backoff + 1u);
+
+		/* Calculate max backoff for the next attempt (~ 2**attempt) */
+		bo->attempt_max_backoff = MIN(bo->attempt_max_backoff * 2u, bo->max_backoff);
+		bo->retries_count++;
+	}
+#else
+	*next_backoff_ms = BACKOFF_CONST_MS;
+#endif
+}
+
+static int aws_client_try_connect(void)
+{
+	int ret;
+	uint32_t backoff_ms;
+	struct backoff_context bo;
+
+	backoff_context_init(&bo);
+
+	while (bo.retries_count <= bo.max_retries) {
+		ret = mqtt_connect(&client_ctx);
+		if (ret == 0) {
+			goto exit;
+		}
+
+		backoff_get_next(&bo, &backoff_ms);
+
+		LOG_ERR("Failed to connect: %d backoff delay: %u ms", ret, backoff_ms);
+		k_msleep(backoff_ms);
+	}
+
+exit:
+	return ret;
+}
+
+struct publish_payload {
+	uint32_t counter;
+};
+
+static const struct json_obj_descr json_descr[] = {
+	JSON_OBJ_DESCR_PRIM(struct publish_payload, counter, JSON_TOK_NUMBER),
+};
+
+static int publish(void)
+{
+	struct publish_payload pl = {.counter = messages_received_counter};
+
+	json_obj_encode_buf(json_descr, ARRAY_SIZE(json_descr), &pl, buffer, sizeof(buffer));
+
+	return publish_message(CONFIG_AWS_PUBLISH_TOPIC, strlen(CONFIG_AWS_PUBLISH_TOPIC), buffer,
+			       strlen(buffer));
+}
+
+void aws_client_loop(void)
+{
+	int rc;
+	int timeout;
+	struct pollfd fds;
+
+	aws_client_setup();
+
+	rc = aws_client_try_connect();
+	if (rc != 0) {
+		goto cleanup;
+	}
+
+	fds.fd = client_ctx.transport.tcp.sock;
+	fds.events = POLLIN | POLLHUP | POLLERR;
```

Comment:
POLLHUP and POLLERR are `revents` only - no need to register for them.

Reply:
Didn't know, fixed.